tools/src/main/java/org/terracotta/utilities/io/Files.java

```diff
@@ -0,0 +1,1856 @@
+import static java.nio.file.StandardCopyOption.COPY_ATTRIBUTES;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Provides <i>sane</i> file management operations.
+ * <p>
+ * Under some operating and file systems -- Windows &amp; HPFS, we're looking at you -- file
+ * operations such as delete and move/rename can be the victim of background or asynchronous
+ * operations.  When, for example, a file is deleted and a directory immediately created in
+ * its place, the directory creation operation can fail because deletion under Windows is
+ * performed asynchronously -- the delete operation simply marks the file system node as
+ * deleted and the actual removal is an asynchronous operation.  When a file is renamed,
+ * that rename can fail because the anti-virus or indexing subsystems might have the file
+ * temporarily open preventing the rename (or move/delete).
+ * <p>
+ * Generally speaking, simply retrying the delete or rename is sufficient to get around the
+ * problem but there may be other complications -- for example, if the rename/move needs
+ * to relocate the file/directory to another file system root a simple retry becomes
+ * potentially burdensome.
+ * <p>
+ * Even though read-only operations, like file copy, should not be affected by background system
+ * tasks, a mirror for {@link java.nio.file.Files#copy(Path, Path, CopyOption...)} is defined in
+ * this class to localize copy code in the event this assertion proves inaccurate and to provide
+ * additional capabilities -- like recursive copying.
+ * <p>
+ * This class does not define a {@code move} method.
+ * Use {@link #relocate(Path, Path, CopyOption...)} relocate} instead.  The {@code relocate}
+ * method performs a rename or copy/delete as necessary.  Recovery of a copy/delete failure
+ * is left as an exercise for the caller.
+ * <p>
+ * Methods in this class retry some operations when presented with an {@link AccessDeniedException}
+ * or one of another {@link FileSystemException} instances indicating potential short-term
+ * interference from other processes.
+ *
+ * <h3>Caution</h3>
+ * Windows, by default, restricts the creation of symbolic links to administrator accounts.
+ * The use of the {@link LinkOption#NOFOLLOW_LINKS NOFOLLOW_LINKS} option will may result
+ * in a {@code FileSystemException} and the copy will fail.  See
+ * <a href="https://stackoverflow.com/a/24353758/1814086">How to create Soft symbolic Link using java.nio.Files</a>
+ * for additional detail.
+ *
+ * @see ExtendedOption
+ * @see <a href="https://docs.microsoft.com/en-us/archive/blogs/junfeng/kb-329065-access-denied-error-when-you-make-code-modifications-with-index-services-running">KB 329065 Access Denied Error When You Make Code Modifications with Index Services Running</a>
+ * @see <a href="https://docs.microsoft.com/en-us/archive/blogs/junfeng/delete-pend-error_access_denied">DELETE PEND == ERROR_ACCESS_DENIED</a>
+ *
+ * @author Clifford W. Johnson
+ */
+public final class Files {
```

Comment:
I think it would be wise to extract out some of this is in to other classes. This thing is pretty enormous and hard to comprehend.  I wonder if we were smart if some of the 'likely to only be used by Windows' stuff could even be moved out in to a class such that we could avoid some of the static initialization costs?

Reply:
Split out about 250 lines.  I neglected to make `FileStoreConstraintException` a top-level class.  The next level would be to remove all except the public methods but I'm not sure that's the right move.